hypervisor/src/mshv/mod.rs

```diff
@@ -1145,7 +1127,62 @@ impl cpu::Vcpu for MshvVcpu {
+                                                ..Default::default()
+                                            };
+                                        self.fd.gpa_read(&mut rax_rw_gpa_arg).map_err(|e| {
+                                            cpu::HypervisorCpuError::GpaRead(e.into())
+                                        })?;
+                                        let data_gpa = u64::from_le_bytes(
+                                            <[u8; 8]>::try_from(&rax_rw_gpa_arg.data[..8]).unwrap(),
+                                        );
+                                        let mut rbx_rw_gpa_arg: mshv_read_write_gpa =
+                                            mshv_bindings::mshv_read_write_gpa {
+                                                base_gpa: ghcb_gpa + GHCB_RBX_OFFSET,
+                                                byte_count: std::mem::size_of::<u64>() as u32,
+                                                ..Default::default()
+                                            };
+                                        self.fd.gpa_read(&mut rbx_rw_gpa_arg).map_err(|e| {
+                                            cpu::HypervisorCpuError::GpaRead(e.into())
+                                        })?;
+                                        let data_npages = u64::from_le_bytes(
+                                            <[u8; 8]>::try_from(&rbx_rw_gpa_arg.data[..8]).unwrap(),
+                                        );
+
+                                        if data_npages > 0 {
+                                            // The certificates are terminated by 24 zero bytes.
+                                            let mut certs_rw_gpa_arg =
+                                                mshv_bindings::mshv_read_write_gpa {
+                                                    base_gpa: data_gpa,
+                                                    byte_count: 16,
+                                                    ..Default::default()
+                                                };
+                                            self.fd.gpa_write(&mut certs_rw_gpa_arg).map_err(
+                                                |e| cpu::HypervisorCpuError::GpaWrite(e.into()),
+                                            )?;
+                                            let mut certs_rw_gpa_arg =
+                                                mshv_bindings::mshv_read_write_gpa {
+                                                    base_gpa: data_gpa + 16,
+                                                    byte_count: 8,
+                                                    ..Default::default()
+                                                };
+                                            self.fd.gpa_write(&mut certs_rw_gpa_arg).map_err(
+                                                |e| cpu::HypervisorCpuError::GpaWrite(e.into()),
+                                            )?;
```

Comment:
Can we refactor this as a generic helper function which takes a buffer and write it into guest by breaking the buffer into 16 bytes chunk? This will simplifying the other places as well where we are writing a buffer into the guest.

Reply:
Sure. I added a helper for reading data as well.

Reply:
Thanks for refactoring the existing code.